lua/nvim_lsp/jdtls.lua

```diff
@@ -0,0 +1,139 @@
+local configs = require 'nvim_lsp/configs'
+local util = require 'nvim_lsp/util'
+local callbacks = require 'vim.lsp.callbacks'
+local path = util.path
+
+local server_name = "jdtls"
+
+local function make_installer()
+  local install_dir = path.join { util.base_install_dir, server_name }
+  local tar_name = "jdt-language-server-latest.tar.gz"
+  local script = string.format([[
+    curl -LO http://download.eclipse.org/jdtls/snapshots/%s > %s
+    tar xf %s
+  ]], tar_name, tar_name, tar_name)
+  local launcher_ls = "ls " .. path.join { install_dir, "plugins", "org.eclipse.equinox.launcher_*.jar" }
+
+  local X = {}
+
+  function X.install()
+    if not util.has_bins("curl", "tar") then
+      error('Need the binaries "curl", "tar" to install this.')
+      return
+    end
+
+    vim.fn.mkdir(install_dir, "p")
+    util.sh(script, install_dir)
+  end
+
+  function X.info()
+    return {
+      is_installed = util.path.exists(install_dir, 'features') ~= false;
+      install_dir = install_dir;
+      launcher = X.get_launcher();
+    }
+  end
+
+  function X.get_os_config()
+    if vim.fn.has("osx") == 1 then
+      return "config_mac"
+    elseif vim.fn.has("unix") == 1 then
+      return "config_linux"
+    else
+      return "config_win"
+    end
+  end
+
+  function X.get_launcher()
+    local file = io.popen(launcher_ls)
+    local results = {}
+
+    for line in file:lines() do
+      table.insert(results, line)
+    end
+
+    if #results == 1 then
+      return results[1]
+    end
+
+    error("Could not find launcher for jdtls.")
+  end
+
+  function X.configure(config)
+    local install_info = X.info()
+    local launcher_path = install_info.launcher
+
+    if install_info.is_installed then
+      config.cmd = vim.list_extend(
+        vim.list_extend(
+          {
+            "java",
+            "-Declipse.application=org.eclipse.jdt.ls.core.id1",
+            "-Dosgi.bundles.defaultStartLevel=4",
+            "-Declipse.product=org.eclipse.jdt.ls.core.product",
+            "-Dlog.level=ALL",
+            "-noverify",
+            "-Xmx1G",
+          },
+          config.init_options.jvm_args),
+        {
+          "-jar", launcher_path,
+          "-configuration", path.join { install_dir, config.init_options.os_config or X.get_os_config() },
+          "-data", config.init_options.workspace,
+          -- TODO: Handle Java versions 8 and under. This may just work...
+          "--add-modules=ALL-SYSTEM",
+          "--add-opens", "java.base/java.util=ALL-UNNAMED",
+          "--add-opens", "java.base/java.lang=ALL-UNNAMED"
+        })
+    end
+  end
+
+  return X
+end
+
+local installer = make_installer()
+
+configs[server_name] = {
+  default_config = {
+    filetypes = { "java" };
+    root_dir = util.root_pattern('.git');
+    init_options = {
+      workspace = vim.loop.os_homedir();
```

Comment:
I don't write Java.
@mfussenegger Do you think this is appropriate?

Reply:
I don't write it "full time" (maintain a REST service layer) but this would be the eclipse workspace... which I'm not sure where that is for people. It can be changed during `setup` but if there is a better default, I'm all for it.

Reply:
According to https://github.com/eclipse/eclipse.jdt.ls/wiki/Running-the-JAVA-LS-server-from-the-command-line the `workspace` option doesn't exist. There would be `workspaceFolders`, but that is optional. 

But the `-data` argument has to point to the workspace folder. Which I guess is done here:  https://github.com/neovim/nvim-lsp/pull/245/files#diff-eaf8d0e99cf1a581b37366966939af30R82

Not sure if `os_homedir` is a good choice. The eclipse.jdt.ls wiki says:

> Choosing a value for -data: the value for your data directory, should be the directory where your active workspace is, and you wish for the java langserver to add in its default files. Should also be the absolute path to this directory, ie., /home/username/workspace


I personally have set it to `~/.local/share/eclipse`, and I even choose a different sub-folder for each project, because I've otherwise had problems when switching between different projects or working on 2 in parallel in the past, but since this is only a default, going with `$HOME/workspace` may be a good enough choice?

Reply:
That sounds good to me. `workspace` is just what i called the option for setup. If you look at the configure method it sets it to `-data`. I feel like putting `data` as an option might be a little to vague.